File: github_cpp/10/117.cpp
class Solution {
public:
	int minDistance(string word1, string word2) {
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    
	    int m = word1.size();
	    int n = word2.size();
	    if (m == 0) { return n; }
	    if (n == 0) { return m; }
	    vector<int> f(n+1);
	    for (int i = 0; i <= n; i++) {
	        f[i] = i;
	    }
	    for (int row = 1; row <= m; row++) {
	        int save = f[0];
	        f[0]++;
	        for (int i = 1; i <= n; i++) {
	            int new_save = f[i];
	            if (word1[row-1] == word2[i-1]) {
	                f[i] = save;
	            }
	            else {
	                f[i] = min(min(save, f[i-1]), f[i]) + 1;
	            }
	            save = new_save;
	        }
	    }
	    return f[n];
	}
};

File: github_code_index/java/8/117.java
public class Solution {
    /**
     * @param word1 and word2: Two string.
     * @return: The minimum number of steps.
     */
    public int minDistance(String a, String b) {
        // write your code 
        
       if (a == null && b == null) {
    		return 0;
    	}

    	int n = a.length();
    	int m = b.length();
    	
    	int[][] f = new int[n + 1][m + 1];	//f[i][j]: the minimal distance of (the first i char of a and the first j of b)

    	for (int i = 0; i < f.length; i++) {
    		f[i][0] = i;
    	}

    	for (int j = 0; j < f[0].length; j++) {
    		f[0][j] = j;
    	}

    	for (int aFirst = 1; aFirst <= a.length(); aFirst++) {
    		for (int bFirst = 1; bFirst <= b.length(); bFirst++) {
    			if (a.charAt(aFirst - 1) == b.charAt(bFirst - 1)) {
    				f[aFirst][bFirst] = f[aFirst - 1][bFirst - 1];
    			} else {    
    				f[aFirst][bFirst] = 1 + Math.min(Math.min(f[aFirst - 1][bFirst - 1], f[aFirst][bFirst - 1]), f[aFirst - 1][bFirst]);
    			}

    		} 
    	}
    	return f[n][m];
    	
    }
}2